framework/Optimizers/GradientDescent.py

```diff
@@ -0,0 +1,840 @@
+      @ In, optVal, float, value of objective variable (corrected for min/max)
+    """
+    self._stepTracker[traj]['grads'].append((rlz, info))
+
+  # * * * * * * * * * * * * * * * *
+  # Resolving potential opt points
+  def _applyBoundaryConstraints(self, point):
+    """
+      Checks and fixes boundary constraints of variables in "point" -> DENORMED point expected!
+      @ In, point, dict, potential point against which to check
+      @ Out, point, dict, adjusted variables
+      @ Out, modded, bool, whether point was modified or not
+    """
+    # TODO should some of this go into the parent Optimizer class, such as the boundary acquiring?
+    modded = False
+    for var in self.toBeSampled:
+      dist = self.distDict[var]
+      val = point[var]
+      lower = dist.lowerBound
+      upper = dist.upperBound
+      if lower is None:
+        lower = -np.inf
+      if upper is None:
+        upper = np.inf
+      if val < lower:
+        self.raiseADebug(' BOUNDARY VIOLATION "{}" suggested value: {:1.3e} lower bound: {:1.3e} under by {:1.3e}'
+                         .format(var, val, lower, lower - val))
+        self.raiseADebug(' ... -> for point {}'.format(point))
+        point[var] = lower
+        modded = True
+      elif val > upper:
+        self.raiseADebug(' BOUNDARY VIOLATION "{}" suggested value: {:1.3e} upper bound: {:1.3e} over by {:1.3e}'
+                         .format(var, val, upper, val - upper))
+        self.raiseADebug(' ... -> for point {}'.format(point))
+        point[var] = upper
+        modded = True
+    return point, modded
+
+  def _applyFunctionalConstraints(self, suggested, previous):
+    """
+      @ In, suggested, dict, NORMALIZED suggested point
+      @ In, previous, dict, NORMALIZED previous point
+      @ Out, suggested, dict, fixed up normalized point
+      @ Out, modded, bool, True if point was modified within this method
+    """
+    # assume no modifications until proved otherwise
+    modded = False
+    # are we violating functional constraints?
+    passFuncs = self._checkFunctionalConstraints(self.denormalizeData(suggested))
+    # while in violation of constraints ...
+    info = {'minStepSize': self._convergenceCriteria.get('stepSize', 1e-10)} # TODO why 1e-10?
+    tries = 500
+    while not passFuncs:
+      modded = True
+      #  try to find new acceptable point
+      denormed = self.denormalizeData(suggested)
+      ### DEBUGG the following lines will add constraint search attempts to the solution export.
+      # rlz = {'trajID': 0,
+      #        'x': denormed['x'],
+      #        'y': denormed['y'],
+      #        'ans': 1 - tries / 100,
+      #        'stepSize': 9999,
+      #        'iteration': 9999,
+      #        'accepted': 'search',
+      #        'conv_gradient': 0,
+      #       }
+      # rlz = dict((key, np.atleast_1d(val)) for key, val in rlz.items())
+      # self._solutionExport.addRealization(rlz)
+      ### END DEBUGG
```

Comment:
This is really useful for debugging behavior, so I humbly recommend leaving it rather than removing the commented code.